trinity/chains/light.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import inspect
+from typing import (  # noqa: F401
+    Any,
+    Optional,
+    Callable,
+    cast,
+    Dict,
+    Generator,
+    Iterator,
+    Tuple,
+    Type,
+    TYPE_CHECKING,
+    Union,
+)
+
+from eth_typing import (
+    BlockNumber,
+    Hash32,
+)
+
+from evm.chains.base import (
+    AccountState,
+    BaseChain,
+)
+from evm.db.backends.base import BaseDB
+from evm.db.chain import (
+    BaseChainDB,
+)
+from evm.db.header import (
+    BaseHeaderDB,
+)
+from evm.exceptions import (
+    HeaderNotFound,
+    VMNotFound,
+)
+from evm.rlp.blocks import (
+    BaseBlock,
+)
+from evm.rlp.headers import (
+    BlockHeader,
+    HeaderParams,
+)
+from evm.rlp.transactions import (
+    BaseTransaction,
+    BaseUnsignedTransaction,
+)
+from evm.validation import (
+    validate_block_number,
+)
+
+from p2p.lightchain import (
+    LightPeerChain,
+)
+
+if TYPE_CHECKING:
+    from evm.vm.base import BaseVM  # noqa: F401
+
+
+class LightDispatchChain(BaseChain):
+    """
+    Provide the :class:`BaseChain` API, even though only a
+    :class:`LightPeerChain` is syncing. Store results locally so that not
+    all requests hit the light peer network.
+    """
+
+    ASYNC_TIMEOUT_SECONDS = 10
+    _loop = None
+
+    def __init__(self, headerdb: BaseHeaderDB, peer_chain: LightPeerChain) -> None:
+        self._headerdb = headerdb
+        self._peer_chain = peer_chain
+        self._peer_chain_loop = asyncio.get_event_loop()
+
+    #
+    # Helpers
+    #
+    @classmethod
+    def get_chaindb_class(cls) -> Type[BaseChainDB]:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    #
+    # Chain API
+    #
+    @classmethod
+    def from_genesis(cls,
+                     base_db: BaseDB,
+                     genesis_params: Dict[str, HeaderParams],
+                     genesis_state: AccountState=None) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    @classmethod
+    def from_genesis_header(cls,
+                            base_db: BaseDB,
+                            genesis_header: BlockHeader) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
```

Comment:
This pattern of implement exception raising methods on required methods from our ABCs is showing up in a few places.  Can we macro this?

```python
class Thing(BaseThing):
    method_a = stub_abstract_method()
```

Should let us drop a lot of code overhead and we don't really care about the type signature for these methods when we do this.

Reply:
So I just realized that doing it this way interferes with the inspection. There are some other options, but I'm not clear how common a case this is. It's really common in `BaseThing`, but we care about the method signature there.

The intention here in `LightDispatchChain` is to implement these methods now, or delete the signatures altogether, and let the parent class raise the warning. I could do something for the superclass, like:

```py
class BaseThing(abc):
    @abstract_runtime_raise
    def method(self, arg):
        pass
```

or

```py
class BaseThing(abc):
    @abstractmethod
    def method(self, arg):
        raise UnimplementedException(self)
```

Reply:
I don't love the decorator approach. As-is is fine for now.  The overhead isn't too great and if we don't have a good solution I'd rather not waste time architecting.